### PR TITLE
Fixes #21449 - Drop host from known hosts on proxies

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

**Comment by tux93** [comment]
I tested this by applying it on a foreman-1.22.1 server and it seems currently the keys are only removed from known_hosts upon deletion but kept when only re-building a host.

To completely solve #21449 I believe the keys need to also be removed when re-building a host.

**Comment by adamruzicka** [comment]
@tux93 thank you for testing this. As far as I understood, with these changes the keys should be removed even on rebuild. Looks like I'll have to look into it a bit more

**Comment by adamruzicka** [comment]
Tests are green-ish

**Comment by tux93** [comment]
I tried again with the current state and keys are still only removed upon deletion, not rebuild, though that could also be due to the old foreman version on my test system

File: app/models/concerns/foreman_remote_execution/orchestration/ssh.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                 end
+      "ssh_remove_known_hosts_#{kind}_#{id}"
+    end
+  end
+
+  module Orchestration::SSHManaged
+    extend ActiveSupport::Concern
+
+    included do
+      register_rebuild(:rebuild_ssh, N_("SSH_#{self.to_s.split('::').first}"))
```
Comment (adamruzicka) [inline app/models/concerns/foreman_remote_execution/orchestration/ssh.rb]: What is this good for? I wasn't able to trigger anything using just this, I had to trigger things from after_validation callback

**Comment by tux93** [comment]
I got around to re-test today and it works now for rebuilding too!

**Comment by ares** [comment]
oh actually not merging, tests failures are related `NameError: undefined local variable or method `queue' for #<Nic::Base:0x000000000beb5180>`

**Comment by adamruzicka** [comment]
[test foreman_remote_execution]

**Comment by adamruzicka** [comment]
@tux93 could you please try this one more time?

**Comment by tux93** [comment]
Re-tested with the latest version of the patch and it still works for me

**Comment by adamruzicka** [comment]
[test foreman_remote_execution]

**Comment by adamruzicka** [comment]
[test foreman_remote_execution]

File: app/models/concerns/foreman_remote_execution/orchestration/ssh.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    def queue_ssh_destroy
+      return unless should_drop_from_known_hosts?
+
+      ssh_destroy
+    end
+
+    def should_drop_from_known_hosts?
+      host, = host_kind_target
+      host && host.build && host.changes.key?('build')
```
[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:42]: this is still on, even though I'm not sure I like the new style more :thinking: 
```suggestion
      host&.build && host&.changes&.key?('build')
```

File: app/models/concerns/foreman_remote_execution/orchestration/ssh.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    def queue_ssh_destroy
+      return unless should_drop_from_known_hosts?
+
+      ssh_destroy
+    end
+
+    def should_drop_from_known_hosts?
+      host, = host_kind_target
```
Comment (ezr-ondrej) [inline app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:41]:
```suggestion
      host, _, _ = host_kind_target
```
seems bit more readable as it states what we are dropping.

File: app/models/concerns/foreman_remote_execution/orchestration/ssh.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    end
+
+    def ssh_destroy
+      logger.debug "Scheduling SSH known_hosts cleanup"
+      queue.create(id: queue_id, name: _("Remove SSH known hosts for %s") % self,
+        priority: 200, action: [self, :drop_from_known_hosts])
+    end
+
+    def queue_ssh_destroy
+      return unless should_drop_from_known_hosts?
```
[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:35]: shouldn't we disable this cop? We've decided to disable it in core.. (theforeman/foreman#7518) this is exactly the case where it doesn't deserve the blank line.

[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:35]: I'm all in for this

File: app/models/concerns/foreman_remote_execution/orchestration/ssh.rb (hunk 3 above)
Comment (adamruzicka) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb]: :police_officer: doesn't like this
```
C: Style/TrailingUnderscoreVariable: Do not use trailing _s in parallel assignment. Prefer host, = host_kind_target.
      host, _, _ = host_kind_target
```

**Comment by adamruzicka** [comment]
We're finally :green_apple:

File: app/models/concerns/foreman_remote_execution/orchestration/ssh.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        return false
+      end
+      true
+    end
+
+    def ssh_destroy
+      logger.debug "Scheduling SSH known_hosts cleanup"
+
+      host, _kind, target = host_kind_target
+      proxies = %w(SSH Ansible).map { |kind| host.remote_execution_proxies(kind).values }
```
Comment (ezr-ondrej) [inline app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:27]: Why do we care about Ansible here? it just sounds wrong :thinking: 
We can't have Ansible without SSH anyway, or am I missing something?
```suggestion
      proxies = host.remote_execution_proxies('SSH').values
```

Comment (adamruzicka) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:27]: Right, we shouldn't have proxies with ansible without ssh. I just wanted to explicitly mention anything that could put records into the know hosts file

Comment (ezr-ondrej) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:27]: hmm understood, it just feels weird as REX should ideally know nada about Ansible, right?

Comment (adamruzicka) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb]: You're right, changed.

File: app/models/concerns/foreman_remote_execution/orchestration/ssh.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      end
+      true
+    end
+
+    def ssh_destroy
+      logger.debug "Scheduling SSH known_hosts cleanup"
+
+      host, _kind, target = host_kind_target
+      proxies = host.remote_execution_proxies('SSH').values
+      proxies.flatten.uniq.each do |proxy|
```
Comment (ezr-ondrej) [inline app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:28]: no need to flatten now, rith?
```suggestion
      host.remote_execution_proxies('SSH').values.each do |proxy|
```

Comment (adamruzicka) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:28]: We cannot, `host.remote_execution_proxies('SSH')` returns a hash of `Symbol -> Array<SmartProxy>`, so at this point, we're operating on `Array<Array<SmartProxy>>` and we need to flatten (and possibly even `uniq`) it

Comment (ezr-ondrej) [reply app/models/concerns/foreman_remote_execution/orchestration/ssh.rb:28]: ah, got it, thaks for expelation :+1:

**Comment by ezr-ondrej** [comment]
WWFJ
hurry-up

**Comment by ezr-ondrej** [comment]
nooo! now we go again! xD I wanted to squash xD

**Comment by adamruzicka** [comment]
I can always dig through reflog and un-squash it if you want it that bad :)

**Comment by ezr-ondrej** [comment]
> I can always dig through reflog and un-squash it if you want it that bad :)

Nah xD It was just about Jenkins being almost done xD and now he goes again :clock1:

**Comment by adamruzicka** [comment]
:green_apple:

**Pull request merged**